### DISCUSSION TITLE
[Bug?] celery.contrib.testing.worker.start_worker's concurrency attribute doesn't seem to work as intended

### DISCUSSION THREAD

**Discussion by SebdeWell** [celery/celery#7366] [Q&A]
# My problem 
I'm trying to write some functional tests for a Django app that uses Celery. Those tests call celery tasks that must be asynchronous and _concurrent_.

In order to be more flexible and run those tests on gitlab pipelines (some times we want to have `app.conf.task_always_eager = True` in our tests to make them synchronous) each [test class](https://docs.djangoproject.com/en/4.0/topics/testing/overview/) will in it's `setUpClass` method start it's own celery worker with 
``` python
from celery.contrib.testing.worker import start_worker
# ...
cls.celery_worker = start_worker(app, log="info", perform_ping_check=False, concurrency=2)
cls.celery_worker.__enter__()         # and __exit__ in tearDownClass() of course
``` 

However, it seems that the `concurrency=2` param is useless, as it doesn't allow for any kind of concurrency.

# To reproduce

Install the the latest celery version (`5.2.3`).

In _celery_tasks.py_:
``` python
import time
from celery import Celery, shared_task

app = Celery("main_app")

app.config_from_object("celery_config")


@shared_task
def sleep_task(secs):
    print(f"Sleeping for {secs} seconds...")
    for i in range(secs):
        time.sleep(i)
        print(f"\t{i + 1}")
    return "DONE"
```

In _celery_config.py_:
``` python
broker_url = "redis://127.0.0.1:6379/0"
result_backend = "redis://127.0.0.1:6379/0"
```

### Normal way to start a task (what is expected):
- in a terminal , use the command `python -m celery -A celery_tasks worker`
- in _main.py_, write and execute in another terminal:
``` python 
from celery_tasks import sleep_task

if __name__ == '__main__':
    sleep_task.delay(3)
    sleep_task.delay(3)
```
Now, in the terminal which runs celery, you can see logs 
```
[2022-03-21 11:42:21,866: WARNING/ForkPoolWorker-7] Sleeping for 3 seconds...
[2022-03-21 11:42:21,866: WARNING/ForkPoolWorker-1] Sleeping for 3 seconds...
[2022-03-21 11:42:21,866: WARNING/ForkPoolWorker-7]     1
[2022-03-21 11:42:21,867: WARNING/ForkPoolWorker-1]     1
[2022-03-21 11:42:22,868: WARNING/ForkPoolWorker-1]     2
[2022-03-21 11:42:22,868: WARNING/ForkPoolWorker-7]     2
[2022-03-21 11:42:24,869: WARNING/ForkPoolWorker-1]     3
[2022-03-21 11:42:24,870: WARNING/ForkPoolWorker-7]     3
``` 
We clearly see that those tasks are asynchronous and _concurrent_

### We start our own worker:
- Do not start a celery worker in another terminal (the script will handle it)
- in _main.py_, write and execute:
``` python
from time import sleep
from celery_tasks import sleep_task, app
from celery.contrib.testing.worker import start_worker

if __name__ == '__main__':
    app.conf.task_always_eager = False        # will sometimes be True, but in that case the concurrence question is irrelevant
    celery_worker = start_worker(app, perform_ping_check=False, concurrency=2)
    with celery_worker:      # equivalent to celery_worker.__enter__() and automatic __exit__()
        print("queuing first sleep_task()")
        sleep_task.delay(3)
        print("queuing second sleep_task()")
        sleep_task.delay(3)
        sleep(10)       # this sleep is here to see full output and allow both tasks to execute
```
Logs produced, in the same terminal where _main.py_ was executed:
```
queuing first sleep_task()
queuing second sleep_task()
Sleeping for 3 seconds...
        1
        2
        3
Sleeping for 3 seconds...
        1
        2
        3
```
As you can see, even when the call were asynchronous, they are not _concurrent_, with both tasks executing one after the other, even with the `concurrency=2` param in `start_worker`

Is this a bug, or is this intended ? If this is intended, what's your suggestion to make it work. To be clear, I want those test classes to run on gitlab pipelines, have some of them behave synchronously with `app.conf.task_always_eager = True`

**Comment by SebdeWell** [comment]
up

**Comment by msdm** [comment]
This is happening because the default pool (`solo`) does not support concurrency. Passing `pool="prefork"` to `start_worker` will make it work.
